server/src/test/java/org/elasticsearch/cluster/health/NoDataNodesHealthTests.java

```diff
@@ -0,0 +1,153 @@
+
+        DiscoveryNodes node = DiscoveryNodes.builder().add(newNode("node_m", Collections.singleton(DiscoveryNode.Role.MASTER))).build();
+        MetaData metaData = MetaData.builder()
+                .put(IndexMetaData.builder("TestIndex")
+                        .settings(settings(Version.CURRENT))
+                        .numberOfShards(randomIntBetween(1, 3))
+                        .numberOfReplicas(randomIntBetween(0, 2)))
+                .build();
+        RoutingTable routingTable = RoutingTable.builder().addAsNew(metaData.index("TestIndex")).build();
+        ClusterState state = ClusterState.builder(new ClusterName("test_cluster"))
+                .nodes(node)
+                .metaData(metaData)
+                .routingTable(routingTable)
+                .build();
+        MockAllocationService service = createAllocationService();
+        state = service.reroute(state, "reroute");
+        return state;
+    }
+
+    public void testClusterHealthWithNoDataNodes() {
+        ClusterState state = setUpClusterWithNoDataNodes();
+        int dataNodes = state.nodes().getDataNodes().size();
+        int masterNodes = state.nodes().getMasterNodes().size();
+        assertTrue(dataNodes == 0);
+        assertTrue(masterNodes > 0);
+        ClusterHealthStatus clusterHealthStatus = new ClusterStateHealth(state).getStatus();
+        assertEquals(ClusterHealthStatus.RED, clusterHealthStatus);
+    }
+
+    /**
+     * The method test for scenario where we have a cluster with indices and data nodes
+     * and then all data nodes gets terminated now new index is created then
+     * all indices and cluster health should be red, but last allocation attempts of new index shards v/s
+     * old index shards should be different.
+     */
+    public void testAllocationStatusForTerminatedNodes() {
+        //creates one master and two data nodes
+        DiscoveryNodes.Builder nodeBuilder = DiscoveryNodes.builder().add(newNode("node_m", Collections.singleton(DiscoveryNode.Role.MASTER)))
+                .add(newNode("node_d1", Collections.singleton(DiscoveryNode.Role.DATA)))
+                .add(newNode("node_d2", Collections.singleton(DiscoveryNode.Role.DATA)));
+        MetaData metaData = MetaData.builder()
+                .put(IndexMetaData.builder("TestIndex")
+                        .settings(settings(Version.CURRENT))
+                        .numberOfShards(2)
+                        .numberOfReplicas(0))
+                .build();
+        RoutingTable routingTable = RoutingTable.builder().addAsNew(metaData.index("TestIndex")).build();
+        ClusterState state = ClusterState.builder(new ClusterName("test_cluster"))
+                .nodes(nodeBuilder.build())
+                .metaData(metaData)
+                .routingTable(routingTable)
+                .build();
+        MockAllocationService allocationService = createAllocationService();
+        //perform allocation of TestIndex
```

Comment:
There's no real need for these comments here, since they only reiterate what the test is doing. I prefer comments only when there's something tricky to explain.